keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -231,15 +238,34 @@ class ApplicationService(
     }
 
     val artifactSummaries = deliveryConfig.artifacts.map { artifact ->
-      val artifactVersionSummaries = repository.artifactVersions(artifact, limit).map { artifactVersion ->
+      val artifactVersions = repository.artifactVersions(artifact, limit)
+
+      // A verification context identifies an artifact version in an environment
+      // For each context, there may be multiple verifications (e.g., test-container, canary)
+      //
+      // This map associates a context with this collection of verifications and their states
+      val vStateMap = getVerificationStates(deliveryConfig, artifactVersions)
+
+      val artifactVersionSummaries = artifactVersions.map { artifactVersion ->
         val artifactSummariesInEnvironments = mutableSetOf<ArtifactSummaryInEnvironment>()
 
         envSummaries.forEach { environmentSummary ->
           val environment = deliveryConfig.environments.find { it.name == environmentSummary.name }!!
+          val verifications = vStateMap[VerificationContext(deliveryConfig, environment, artifactVersion)]
```

Comment:
I think this would be slightly more readable if you pulled out the value used as a key:
```suggestion
          val verificationContext = VerificationContext(deliveryConfig, environment, artifactVersion)
          val verifications = vStateMap[verificationContext]
```